src/object/srv_obj.c

```diff
@@ -2978,3 +3023,837 @@ obj_verify_bio_csum(crt_rpc_t *rpc, daos_iod_t *iods,
 
 	return rc;
 }
+
+static inline void
+ds_obj_cpd_set_sub_result(struct obj_cpd_out *oco, int idx,
+			  int result, daos_epoch_t epoch)
+{
+	uint64_t	*p_epoch;
+	int		*p_ret;
+
+	p_epoch = (uint64_t *)oco->oco_sub_epochs.ca_arrays + idx;
+	*p_epoch = epoch;
+
+	p_ret = (int *)oco->oco_sub_rets.ca_arrays + idx;
+	*p_ret = result;
+}
+
+static void
+obj_cpd_reply(crt_rpc_t *rpc, int status, uint32_t map_version)
+{
+	struct obj_cpd_out	*oco = crt_reply_get(rpc);
+	int			 rc;
+
+	obj_reply_set_status(rpc, status);
+	obj_reply_map_version_set(rpc, map_version);
+
+	D_DEBUG(DB_TRACE, "CPD rpc %p send reply, pmv %d, status %d.\n",
+		rpc, map_version, status);
+
+	rc = crt_reply_send(rpc);
+	if (rc != 0)
+		D_ERROR("Send CPD reply failed: "DF_RC"\n", DP_RC(rc));
+
+	if (oco->oco_sub_rets.ca_count != 0) {
+		D_FREE(oco->oco_sub_rets.ca_arrays);
+		oco->oco_sub_rets.ca_count = 0;
+	}
+
+	if (oco->oco_sub_epochs.ca_count != 0) {
+		D_FREE(oco->oco_sub_epochs.ca_arrays);
+		oco->oco_sub_epochs.ca_count = 0;
+	}
+}
+
+/* Locally process the operations belong to one DTX.
+ * Common logic, shared by both leader and non-leader.
+ */
+static int
+ds_obj_dtx_handle_one(crt_rpc_t *rpc, struct daos_cpd_sub_head *dcsh,
+		      struct daos_cpd_disp_ent *dcde,
+		      struct daos_cpd_sub_req *dcsrs,
+		      struct obj_io_context *ioc, struct dtx_handle *dth)
+{
+	struct daos_cpd_req_idx		 *dcri = dcde->dcde_reqs;
+	struct daos_cpd_sub_req		 *dcsr;
+	struct daos_cpd_update		 *dcu;
+	daos_handle_t			 *iohs = NULL;
+	struct bio_desc			**biods = NULL;
+	struct obj_bulk_args		 *bulks = NULL;
+	struct bio_sglist		**bsgls_dups = NULL;
+	struct dcs_iod_csums		 *csums;
+	daos_iod_t			 *iods;
+	uint64_t			 *offs = NULL;
+	int				  rma = 0;
+	int				  rma_idx = 0;
+	int				  rc = 0;
+	int				  i;
+	int				  j;
+
+	/* P1: Spread read TS. */
+	for (i = 0; i < dcde->dcde_read_cnt; i++) {
+		daos_handle_t	ioh;
+
+		dcsr = &dcsrs[dcri[i].dcri_req_idx];
+		if (dcsr->dcsr_opc != DCSO_READ) {
+			D_ERROR(DF_DTI" expected sub read, but got opc %u\n",
+				DP_DTI(&dcsh->dcsh_xid), dcsr->dcsr_opc);
+
+			D_GOTO(out, rc = -DER_PROTO);
+		}
+
+		dcsr->dcsr_oid.id_shard = dcri[i].dcri_shard_idx;
+		rc = vos_fetch_begin(ioc->ioc_coc->sc_hdl, dcsr->dcsr_oid,
+				     dcsh->dcsh_epoch.oe_value, 0,
+				     &dcsr->dcsr_dkey, dcsr->dcsr_nr,
+				     dcsr->dcsr_read.dcr_iods,
+				     VOS_FETCH_SET_TS_ONLY, NULL, &ioh, dth);
+		if (rc == 0)
+			rc = vos_fetch_end(ioh, 0);
+		else if (rc == -DER_NONEXIST)
+			rc = 0;
+
+		if (rc != 0) {
+			D_ERROR("Failed to set read TS for obj "DF_UOID
+				", DTX "DF_DTI": "DF_RC"\n",
+				DP_UOID(dcsr->dcsr_oid),
+				DP_DTI(&dcsh->dcsh_xid), DP_RC(rc));
+			goto out;
+		}
+	}
+
+	dcri += dcde->dcde_read_cnt;
+	/* P2: vos_update_begin. */
+	for (i = 0; i < dcde->dcde_write_cnt; i++) {
+		dcsr = &dcsrs[dcri[i].dcri_req_idx];
+		dcsr->dcsr_oid.id_shard = dcri[i].dcri_shard_idx;
+
+		if (dcsr->dcsr_opc != DCSO_UPDATE)
```

Comment:
Just curious, it only support update for CPD now? or it will support more later?

Reply:
No, we support others. The logic here just only for filter out update sub-requests.